tests/Agency/StorePerformanceTest.cpp

```diff
@@ -0,0 +1,651 @@
+  }
+  op.report();
+}
+
+// test array operations specifically
+
+TEST_F(StorePerformanceTest, array_push) {
+  writeAndCheck(R"([[{"/a/b/c":[]}]])");
+  std::vector<consensus::query_t> write_queries;
+  size_t i {};
+  std::generate_n(std::back_inserter(write_queries), repetition_times[1], [&i](){
+    return VPackParser::fromJson(R"([[{"/a/b/c":{"op":"push","new":)" + std::to_string(i++) + "}}]]");
+  });
+  OperationMeasurement op{repetition_times[1]};
+  for (auto const& wq: write_queries) {
+    writeAndCheck(wq);
+    ++op;
+  }
+  op.report();
+}
+
+TEST_F(StorePerformanceTest, array_pop) {
+  std::stringstream ss;
+  ss << R"([[{"/a/b/c":[0)";
+  for (size_t i{1}; i < repetition_times[1]; ++i) {
+    ss << "," << i;
+  }
+  ss << "]}]]";
+  writeAndCheck(ss.str());
+  auto const rq {VPackParser::fromJson(R"([[{"/a/b/c":{"op":"pop"}}]])")};
+  OperationMeasurement op{repetition_times[1]};
+  for (size_t i{}; i < repetition_times[1]; ++i) {
+    writeAndCheck(rq);
+    ++op;
+  }
+  op.report();
+}
+
+// test object operations specifically
+
+// test operations which need to change a lot in the tree
+// add in ascending order
+TEST_F(StorePerformanceTest, tree_add_ascending) {
+  std::vector<consensus::query_t> write_queries;
+  auto const width {log10(repetition_times[3])};
+  for (size_t i{}; i < repetition_times[3]; ++i) {
+    auto k {std::to_string(i)};
+    k = std::string(width - k.size(), '0') + k;
+    write_queries.push_back(VPackParser::fromJson("[[{\"k"+k+"\": 42}]]"));
+  }
+  OperationMeasurement op{repetition_times[3]};
+  for (auto const& wq: write_queries) {
+    write(wq);
+    ++op;
+  }
+  op.report();
+}
+
+// add in descending order
+TEST_F(StorePerformanceTest, tree_add_descending) {
+  std::vector<consensus::query_t> write_queries;
+  auto const width {log10(repetition_times[3])};
+  for (auto i{repetition_times[3]}; i > 0; --i) {
+    auto k {std::to_string(i)};
+    k = std::string(width - k.size(), '0') + k;
+    write_queries.push_back(VPackParser::fromJson("[[{\"k"+k+"\": 42}]]"));
+  }
+  OperationMeasurement op{repetition_times[3]};
+  for (auto const& wq: write_queries) {
+    write(wq);
+    ++op;
+  }
+  op.report();
+
+
+}
+
+// add ascending, remove root, then re-add
+TEST_F(StorePerformanceTest, tree_add_remove_readd) {
+  std::vector<consensus::query_t> write_queries;
+  auto const width {log10(repetition_times[3])};
+  for (size_t i{}; i < repetition_times[3]; ++i) {
+    auto k {std::to_string(i)};
+    k = std::string(width - k.size(), '0') + k;
+    write_queries.push_back(VPackParser::fromJson("[[{\"k"+k+"\": 42}]]"));
+  }
+  OperationMeasurement op{repetition_times[3]};
+  for (auto const& wq: write_queries) {
+    write(wq);
+    ++op;
+  }
+  write("[[{\"/t\": 0}]]");
+  for (auto const& wq: write_queries) {
+    write(wq);
+    ++op;
+  }
+  op.report();
+}
+
+// test for contention:
+// multiple threads manipulate values
+TEST_F(StorePerformanceTest, multiple_threads_all_separate_keys) {
```

Comment:
We probably want to vary a bit the number of threads and maybe try with 1, 2, 4, 8, 16 and 32 threads to see contention building up (or not).

Reply:
In that case you should also consider thread numbers that are not a power of two - sometimes this generates some surprising results.

Reply:
Sure, can do.